src/test/scala/chiselTests/AsTypeOfTester.scala

```diff
@@ -120,34 +122,559 @@ class AsChiselEnumTester extends BasicTester {
   stop()
 }
 
-class AsTypeOfSpec extends ChiselFlatSpec {
-  behavior.of("asTypeOf")
+object AsTypeOfSpec {
+  class MyBundle extends Bundle {
+    val a = UInt(4.W)
+    val b = UInt(4.W)
+  }
+  class MyBundleUnknownWidth(w: Int) extends Bundle {
+    val a = UInt(w.W)
+    val b = UInt()
+  }
+  object MyEnum extends ChiselEnum {
+    val a, b = Value
+    val c = Value(10.U)
+  }
+
+  class SimpleAsTypeOf[A <: Data, B <: Data](gen1: A, gen2: B) extends RawModule {
+    val in = IO(Input(gen1))
+    // Out is a wire so we can use inferred widths
+    val out = dontTouch(Wire(gen2))
+    out :#= in.asTypeOf(out)
+  }
+  class WireSourceAsTypeOf[A <: Data, B <: Data](gen1: A, gen2: B)(driveIn: A => Unit) extends RawModule {
+    // Called in so that tests can check for same Verilog from this and SimpleAsTypeOf
+    val in = dontTouch(Wire(gen1))
+    driveIn(in)
+    val out = dontTouch(Wire(gen2))
+    out := in.asTypeOf(out)
+  }
+
+  def emitSystemVerilog(design: => RawModule): String = {
+    ChiselStage.emitSystemVerilog(design)
+  }
+}
+
+class AsTypeOfSpec extends ChiselFunSpec {
+  import AsTypeOfSpec._
+
+  describe("asTypeOf") {
+
+    it("should work with Bundles containing Bits Types") {
+      assertTesterPasses { new AsTypeOfBundleTester }
+    }
 
-  it should "work with Bundles containing Bits Types" in {
-    assertTesterPasses { new AsTypeOfBundleTester }
+    it("should work with Bundles that have fields of zero width") {
+      assertTesterPasses { new AsTypeOfBundleZeroWidthTester }
+    }
+
+    it("should work with Vecs containing Bits Types") {
+      assertTesterPasses { new AsTypeOfVecTester }
+    }
+
+    it("should expand and truncate UInts of different width") {
+      assertTesterPasses { new AsTypeOfTruncationTester }
+    }
+
+    it("should work for casting implicit Reset to Bool") {
+      assertTesterPasses { new ResetAsTypeOfBoolTester }
+    }
+
+    it("should work for casting to and from ChiselEnums") {
+      assertTesterPasses(new AsChiselEnumTester)
+    }
+
+    it("should work for casting to and from Clock") {
+      assertTesterPasses(new AsTypeOfClockTester)
+    }
+  }
```

Comment:
This part of the diff is a little weird. These are the existing execution tests that I grouped under a new level of `describe` before adding all of the tests starting at `describe("Bundles") {`.